src/types.rs

```diff
@@ -58,6 +59,137 @@ pub struct SidechainProposal {
+        const HASH_ID_2_LENGTH: usize = 20;
+
+        let description_length = input.len() - HASH_ID_1_LENGTH - HASH_ID_2_LENGTH;
+        let (input, description) = take(description_length)(input)?;
+
+        let (input, hash_id_1) = take(HASH_ID_1_LENGTH)(input)?;
+        let (input, hash_id_2) = take(HASH_ID_2_LENGTH)(input)?;
+
+        let hash_id_1: [u8; HASH_ID_1_LENGTH] = hash_id_1.try_into().unwrap();
+        let hash_id_2: [u8; HASH_ID_2_LENGTH] = hash_id_2.try_into().unwrap();
+
+        let parsed = UnvalidatedDeserializedSidechainProposalV1 {
+            title: title.to_vec(),
+            description: description.to_vec(),
+            hash_id_1,
+            hash_id_2,
+        };
+
+        Ok((input, (sidechain_number, parsed)))
+    }
+
+    pub fn try_deserialize(&self) -> Result<(SidechainNumber, DeserializedSidechainProposalV1)> {
+        let (input, (sidechain_number, unvalidated)) =
+            self.try_deserialize_unvalidated()
+                .map_err(|err| match err {
+                    nom::Err::Error(nom::error::Error {
+                        code: nom::error::ErrorKind::Tag,
+                        input,
+                    }) => SidechainProposalError::UnknownVersion(input[0]),
+                    _ => SidechainProposalError::FailedToDeserialize(err.to_owned()),
+                })?;
+
+        // One might think "this would be a good place to check there's no trailing bytes".
+        // That doesn't work! Our parsing logic consumes _all_ bytes, because the length of the
+        // `description` field is defined as the total length of the entire input minus
+        // the length of all the other, known-length fields. We therefore _always_ read
+        // the entire input.
+        if !input.is_empty() {
+            panic!("somehow ended up with trailing bytes")
+        }
+
+        let sidechain_number = SidechainNumber::from(sidechain_number);
+
+        let title = String::from_utf8(unvalidated.title)
+            .map_err(SidechainProposalError::InvalidUtf8Title)?;
+
+        let description = String::from_utf8(unvalidated.description)
+            .map_err(SidechainProposalError::InvalidUtf8Description)?;
+
+        Ok((
+            sidechain_number,
+            DeserializedSidechainProposalV1 {
+                title,
+                description,
+                hash_id_1: unvalidated.hash_id_1,
+                hash_id_2: unvalidated.hash_id_2,
+            },
+        ))
+    }
+}
+
+/// We know how to deserialize M1 v1 messages from BIP300
+/// https://github.com/bitcoin/bips/blob/master/bip-0300.mediawiki#m1----propose-sidechain
+///
+///    1-byte nSidechain
+///    4-byte nVersion
+///    1-byte title length
+///    x-byte title
+///    x-byte description
+///    32-byte hashID1
+///    20-byte hashID2
+///
+/// Description length is total_data_length - length_of_all_other_fields
+#[derive(Debug)]
+pub struct DeserializedSidechainProposalV1 {
```

Comment:
Perhaps you should implement
*  `TryFrom<SidechainProposal>` for `UnvalidatedDeserializedSidechainProposalV1`
* `From<UnvalidatedDeserializedSidechainProposalV1>` for `SidechainProposal`
* `TryFrom<UnvalidatedDeserializedSidechainProposalV1>` for `DeserializedSidechainProposalV1`
* `From<DeserializedSidechainProposalV1>` for `UnvalidatedDeserializedSidechainProposalV1`
*  `From<DeserializedSidechainProposalV1>` for `SidechainProposal`
* `TryFrom<SidechainProposal>` for `DeserializedSidechainProposalV1`
* `bitcoin::consensus::Decodable` for `UnvalidatedDeserializedSidechainProposalV1`
* `bitcoin::consensus::Decodable` for `DeserializedSidechainProposalV1`
*  `bitcoin::consensus::Encodable` for `UnvalidatedDeserializedSidechainProposalV1`
*  `bitcoin::consensus::Encodable` for `DeserializedSidechainProposalV1`